packages/host/app/services/matrix-service.ts

```diff
@@ -755,8 +756,15 @@ export default class MatrixService extends Service {
       }
     }
 
+    // Upload skill cards and attached cards together to ensure:
+    // 1. Latest skill cards are deployed with the message
+    // 2. FileDefManager's cache prevents re-uploading unchanged skill cards
+    let skillCards = this.roomResources.get(roomId)?.skillCards ?? [];
+    let attachedCardsAndSkills = [...attachedCards, ...skillCards].filter(
+      (card, index, self) => index === self.findIndex((c) => c.id === card.id),
+    );
```

Comment:
I assume this filter logic is supposed to enforce uniqueness. Let's make a small util named accordingly so that intent is clear.